src/utils/size-observer.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {pushIfNotExist, removeItem} from './array';
+import {rethrowAsync} from '../log';
+
+/** @typedef {function(../layout-rect.LayoutSizeDef)} */
+let ContentSizeObserverCallbackDef;
+
+/** @const {!WeakMap<!Window, !ResizeObserver>} */
+const observers = new WeakMap();
+
+/** @const {!WeakMap<!Element, !ContentSizeObserverCallbackDef>} */
+const targetObserverListMap = new WeakMap();
+
+/** @const {!WeakMap<!Element, !../layout-rect.LayoutSizeDef>} */
+const targetSizeListMap = new WeakMap();
+
+/**
+ * @param {!Element} element
+ * @param {!ContentSizeObserverCallbackDef} callback
+ */
+export function observeContentSize(element, callback) {
+  const win = element.ownerDocument.defaultView;
+  if (!win) {
+    return;
+  }
+  let callbacks = targetObserverListMap.get(element);
+  if (!callbacks) {
+    callbacks = [];
+    targetObserverListMap.set(element, callbacks);
+    getObserver(win).observe(element);
+  }
+  if (pushIfNotExist(callbacks, callback)) {
```

Comment:
Do you think `pushIfNotExist` is better than a fail-fast approach (throw if callback is already registered)? I think I agree with the safer approach you've chosen

Reply:
I figured to keep it close to the `ResizeObserver` where calling `observe(element)` twice would do nothing. We approached it a bit differently for in-viewport, but that was because we didn't allow multi-map there.